packages/sol-doc/src/solidity_doc_generator.ts

```diff
@@ -0,0 +1,284 @@
+}
+
+function _genMethodDoc(
+    abiDefinition: MethodAbi | FallbackAbi,
+    devdocIfExists: DevdocOutput | undefined,
+): SolidityMethod {
+    const name = abiDefinition.type === 'fallback' ? '' : abiDefinition.name;
+
+    const { parameters, methodSignature } =
+        abiDefinition.type === 'fallback'
+            ? { parameters: [], methodSignature: `${name}()` }
+            : _genMethodParamsDoc(name, abiDefinition.inputs, devdocIfExists);
+
+    const comment = _devdocMethodDetailsIfExist(methodSignature, devdocIfExists);
+
+    const returnType =
+        abiDefinition.type === 'fallback'
+            ? { name: '', typeDocType: TypeDocTypes.Intrinsic }
+            : _genMethodReturnTypeDoc(abiDefinition.outputs, methodSignature, devdocIfExists);
+
+    const isConstant = abiDefinition.type === 'fallback' ? true : abiDefinition.constant;
+    // TODO: determine whether fallback functions are always constant, as coded just above
+
+    const methodDoc: SolidityMethod = {
+        isConstructor: false,
+        name,
+        callPath: '', // TODO: wtf is this?
+        parameters,
+        returnType,
+        isConstant,
+        isPayable: abiDefinition.payable,
+        comment,
+    };
+    return methodDoc;
+}
+
+function _genEventDoc(abiDefinition: EventAbi): Event {
+    const eventDoc: Event = {
+        name: abiDefinition.name,
+        eventArgs: _genEventArgsDoc(abiDefinition.inputs, undefined),
+    };
+    return eventDoc;
+}
+
+function _genEventArgsDoc(args: DataItem[], devdocIfExists: DevdocOutput | undefined): EventArg[] {
+    const eventArgsDoc: EventArg[] = [];
+
+    for (const arg of args) {
+        const name = arg.name;
+
+        const type: Type = {
+            name: arg.type,
+            typeDocType: TypeDocTypes.Intrinsic,
+        };
+
+        const eventArgDoc: EventArg = {
+            isIndexed: false, // TODO: wtf is this?
+            name,
+            type,
+        };
+
+        eventArgsDoc.push(eventArgDoc);
+    }
+    return eventArgsDoc;
+}
+
+/**
+ * Extract documentation for each method paramater from @param params.
+ */
+function _genMethodParamsDoc(
+    name: string,
+    abiParams: DataItem[],
+    devdocIfExists: DevdocOutput | undefined,
+): { parameters: Parameter[]; methodSignature: string } {
+    const parameters: Parameter[] = [];
+    let methodSignature = `${name}(`;
+
+    for (const abiParam of abiParams) {
+        const parameter: Parameter = {
+            name: abiParam.name,
+            comment: '',
+            isOptional: false, // Unsupported in Solidity, until resolution of https://github.com/ethereum/solidity/issues/232
+            type: { name: abiParam.type, typeDocType: TypeDocTypes.Intrinsic },
+        };
+        parameters.push(parameter);
+        methodSignature = `${methodSignature}${abiParam.type},`;
+    }
+
+    if (methodSignature.slice(-1) === ',') {
+        methodSignature = methodSignature.slice(0, -1);
+    }
+    methodSignature += ')';
```

Comment:
We try to avoid string concatenation. I could have written this as:

```
// Aggregate param types into `paramTypes` array.
const methodSignature = `${name}(${paramTypes.join(',')})`;
```

Reply:
Love it, thank you.